instrumentation/kamon-cats-io/src/test/scala/kamon/instrumentation/cats/io/AbstractCatsEffectInstrumentationSpec.scala

```diff
@@ -0,0 +1,90 @@
+package kamon.instrumentation.cats.io
+
+import java.util.concurrent.Executors
+
+import cats.effect.{Async, ContextShift, Effect, LiftIO, Timer}
+import cats.implicits._
+import kamon.Kamon
+import kamon.context.Context
+import kamon.instrumentation.cats.io.Tracing.Implicits._
+import kamon.tag.Lookups.plain
+import kamon.testkit.TestSpanReporter
+import org.scalatest.concurrent.{PatienceConfiguration, ScalaFutures}
+import org.scalatest.{BeforeAndAfterAll, Inspectors, Matchers, WordSpec}
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContext.global
+import scala.concurrent.duration.Duration
+
+// NOTE: We have this test just to ensure that the Context propagation is working, but starting with Kamon 2.0 there
+//       is no need to have explicit Runnable/Callable instrumentation because the instrumentation brought by the
+//       kamon-executors module should take care of all non-JDK Runnable/Callable implementations.
+abstract class AbstractCatsEffectInstrumentationSpec[F[_]: LiftIO](effectName: String)(implicit F: Effect[F])
+  extends WordSpec
+    with ScalaFutures
+    with Matchers
+    with PatienceConfiguration
+    with TestSpanReporter
+    with Inspectors
+    with BeforeAndAfterAll {
+
+  implicit def contextShift: ContextShift[F]
+
+  implicit def timer: Timer[F]
+
+  private val customExecutionContext = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())
+
+  override protected def afterAll(): Unit = {
+    customExecutionContext.shutdown()
+    shutdownTestSpanReporter()
+    super.afterAll()
+  }
+
+  s"A Cats Effect $effectName" should {
+    "capture the active span available when created" which {
+      "must be available across asynchronous boundaries" in {
+        val context = Context.of("key", "value")
+
+        val contextTagF: F[String] =
+          for {
+            scope    <- F.delay(Kamon.storeContext(context))
+            _        <- Async.shift[F](customExecutionContext)
+            len      <- F.delay("Hello Kamon!").map(_.length)
+            _        <- F.pure(len.toString)
+            _        <- timer.sleep(Duration.Zero)
+            _        <- Async.shift[F](global)
+            tagValue <- F.delay(Kamon.currentContext().getTag(plain("key")))
+            _        <- F.delay(scope.close())
+          } yield tagValue
+
+        val contextTag = F.toIO(contextTagF).unsafeRunSync()
+        contextTag shouldEqual "value"
+      }
+    }
+
+    "nest spans correctly" in {
+      // the test expects the following span tree, but for some reason, it doesn't work:
+      // - root
+      //    - 1 (value = 1)
+      //    - 2 (value = 2)
+      //    - 3 (value = 3)
+      val rootSpan = for {
+        root  <- F.delay(Kamon.spanBuilder("root").start())
+        _     <- (1L to 3L)
```

Comment:
Why would the spans be nested in a tree? 
I thinks this test just creates unrelated spans. 
Maybe try creating the other spans as children of the first one, or using `Kamon.runWithSpan`

Reply:
In the test all operations are sequenced as follows:

- a root span is created and started without being finished
- 3 additional spans that are started and finished, executed one after another (`.map(...).sequence`)
- the root span is finished

The `named` function that is applied to the `F.delay(idx)` is actually a call to the `operationName` in the `Tracing` object that does the following:

- creates and starts a new span with the provided name and tags that sets as parent span the result of `Kamon.currentSpan()`
- executes the `F[_]` effect, which in this case is `F.delay(idx)`
- finishes the span

Assuming the operation sequence defined above, the three effects `F.delay(idx)` should be wrapped in spans that use the `root` span as a parent, because that's the current span (the started and not finished).

I don't understand what part of this test is "iffy". :) ... but please let me know if I'm missing something.

The assertion at the end is obviously wrong because it also considers the root span for the equality check, but that's not the problem here as I've manually printed the span ids when they are created.

Reply:
Looking at the kamon-scala-future module, it looks like it doesn't just create a span, but explicitly stores it in the context storage and removes it afterwards. I assumed that simply starting and finishing a span is sufficient. I'll try to do the same.

Reply:
Hmm, I've put down some println's and `Kamon.currentSpan()` give `Span.Empty` both in the span creation code and in the test code.
Adding something like `Kamon.runWithSpan(root)` to the test should correctly nest it in the test code.

Reply:
Ah, you figured it out! :tada:

Reply:
That was the problem, I needed to store the span in the context storage...
But now I've hit another issue where every now and again the test fails due to the fact that the fibers are interleaved. I think this is a more fundamental issue with my PR, I may have to use TaskLocal and IOLocal to get named spans working correctly. Still investigating.